python/node_similarity.py

```diff
@@ -0,0 +1,304 @@
+    :type node1: mgp.Vertex
+    :param node2: The second node
+    :type node2: mgp.Vertex
+
+    :return: The cosine similarity of the first and the second node
+    :rtype: float
+    """
+
+    cosine_similarity: float
+
+    neighbors1: set
+    neighbors2: set
+
+    if node1 == node2:
+        return 1.0
+
+    cosine_similarity = 0.0
+
+    neighbors1 = _get_neighbors(node1)
+    neighbors2 = _get_neighbors(node2)
+
+    denominator = sqrt(len(neighbors1) * len(neighbors2))
+
+    if denominator != 0:
+        cosine_similarity = len(neighbors1 & neighbors2) / denominator
+
+    return cosine_similarity
+
+
+def _calculate_similarity(
+    node1: Union[mgp.Vertex, Tuple[mgp.Vertex]],
+    node2: Union[mgp.Vertex, Tuple[mgp.Vertex]],
+    method: Callable,
+    mode: Mode,
+) -> List[Tuple[mgp.Vertex, mgp.Vertex, float]]:
+    """This method calculates the similarity of nodes for given method and mode.
+
+    :param node1: The first node or tuple of nodes
+    :type node1: mgp.Vertex, tuple
+    :param node2: The second node or tuple of nodes
+    :type node2: mgp.Vertex, tuple
+    :param method: Similarity measure which will be used for calculating the similarity between nodes.
+    Currently available are `_calculate_jaccard`, `_calculate_overlap` and `_calculate_cosine`
+    :type method: function
+    :param mode: Can be PAIRWISE for pairwise similarity calculation or CARTESIAN
+    for calculating the similarity of the Cartesian product of given tuples.
+    :type mode: enum
+
+    :raises TypeError: Occurs if there's a type mismatch for passed arguments
+    :raises ValueError: Occurs if any passed argument is invalid
+
+    :return: Returns the calculated similarity between nodes in a list of mgp.Records. Each mgp.Record consist of
+    the first node, the second node and the similarity between them.
+    :rtype: list
+    """
+
+    # clear cache
+    neighbors_dict.clear()
+
+    nodes1: tuple
+    nodes2: tuple
+
+    if isinstance(node1, mgp.Vertex):
+        nodes1 = tuple([node1])
+    elif isinstance(node1, tuple):
+        nodes1 = node1
+    else:
+        raise TypeError("Invalid type of first argument.")
+
+    if isinstance(node2, mgp.Vertex):
+        nodes2 = tuple([node2])
+    elif isinstance(node2, tuple):
+        nodes2 = node2
+    else:
+        raise TypeError("Invalid type of second argument.")
+
+    if mode == Mode.PAIRWISE:
+        if len(nodes1) == len(nodes2):
+            return [
+                mgp.Record(node1=n1, node2=n2, similarity=method(n1, n2))
+                for n1, n2 in zip(nodes1, nodes2)
+            ]
+        else:
+            raise ValueError("Incompatible lengths of given arguments.")
+    elif mode == Mode.CARTESIAN:
+        return [
+            mgp.Record(node1=n1, node2=n2, similarity=method(n1, n2))
+            for n1, n2 in product(nodes1, nodes2)
+        ]
+    else:
+        raise ValueError("Invalid mode. Valid modes are: pairwise, cartesian.")
```

Comment:
This can be updated with according to this comment:
https://memgraph.slack.com/archives/C01L7JJL1L5/p1628255422010800